internal/core/domain/trade_service.go

```diff
@@ -4,76 +4,80 @@ import (
 	"time"
 
 	"github.com/shopspring/decimal"
-	"github.com/tdex-network/tdex-daemon/pkg/transactionutil"
-
-	"github.com/tdex-network/tdex-daemon/config"
-	pkgswap "github.com/tdex-network/tdex-daemon/pkg/swap"
-	"github.com/tdex-network/tdex-daemon/pkg/wallet"
-	pb "github.com/tdex-network/tdex-protobuf/generated/go/swap"
-	"google.golang.org/protobuf/proto"
 )
 
-// Propose returns a new trade proposal for the given trader and market
+// Propose brings an Empty trade to the Propose status by first validating the
+// provided arguments.
 func (t *Trade) Propose(
-	swapRequest *pb.SwapRequest,
+	swapRequest SwapRequest,
 	marketQuoteAsset string,
 	marketFeeBasisPoint int64,
+	expiryDuration uint64,
 	traderPubkey []byte,
 ) (bool, error) {
-	if !t.IsEmpty() {
-		return false, ErrMustBeEmpty
+	if t.Status.Code >= Proposal {
+		return true, nil
 	}
 
+	now := uint64(time.Now().Unix())
 	t.TraderPubkey = traderPubkey
 	t.MarketQuoteAsset = marketQuoteAsset
 	t.SwapRequest.ID = swapRequest.GetId()
-	t.Timestamp.Request = uint64(time.Now().Unix())
-	t.Timestamp.Expiry = t.Timestamp.Request +
-		uint64(config.GetInt(config.TradeExpiryTimeKey))
+	t.SwapRequest.Timestamp = now
 	t.PsetBase64 = swapRequest.GetTransaction()
+	t.Status = ProposalStatus
 
-	msg, err := pkgswap.ParseSwapRequest(swapRequest)
+	msg, err := SwapParser.SerializeRequest(swapRequest)
 	if err != nil {
 		t.Fail(
 			swapRequest.GetId(),
-			ProposalRejectedStatus,
-			pkgswap.ErrCodeInvalidSwapRequest,
+			err.Code,
 			err.Error(),
 		)
 		return false, nil
 	}
 
 	price := calculateMarketPrices(swapRequest, marketQuoteAsset)
 
-	t.Status = ProposalStatus
+	t.ExpiryTime = now + expiryDuration
 	t.SwapRequest.Message = msg
 	t.MarketPrice = price
 	t.MarketFee = marketFeeBasisPoint
 	return true, nil
 }
 
-// Accept attempts to accept a trade proposal. The trade must be in Proposal
-// status to be accepted, otherwise an error is thrown
+// Accept brings a trade from the Proposal to the Accepted status by validating
+// the provided arguemtn against the the SwapRequest message.
 func (t *Trade) Accept(
 	psetBase64 string,
 	inputBlindingKeys,
 	outputBlindingKeys map[string][]byte,
 ) (bool, error) {
+	if t.Status.Code >= Accepted {
+		return true, nil
+	}
+
+	if t.IsExpired() {
```

Comment:
Not sure about this: the exiration is "attached" after the swap is accepted by the daemon.

Reply:
A SwapRequest from a client has no expiration: as long the pricing terms are good, we move forward

Reply:
hmmm.. actually, the expiration time was set for the trade within the Proposal method even before this PR.
I'm going to fix this then.